[CORE-13043] cluster_link: enable changing bootstrap servers with update shadow link command

IoannisRP opened the pull request:
Implements: [CORE_13043](https://redpandadata.atlassian.net/browse/CORE-13043)

Add path to handle changes in bootstrap_servers and/or tls config of a cluster link.

When either one of these change as part of a shadow link update request, the brokers in the `kafka::client::cluster` need to be brought down and be remade using the new seeds/tls config.

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes

* none

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

IoannisRP committed:
dt: add test to verify cluster link broker update

Because of the difficulty involved in changing the address of a cluster
in dt, this test is a workaround. The second source cluster is meant to
represent a change of broker address.

IoannisRP committed:
r/a/s/shadow_link: allow updating bootstrap and tls_config settings

IoannisRP committed:
cluster_link: fix type in comment

IoannisRP committed:
k/client: Extend update config to update everything in kafka client cluster

This commit also add all the functionality needed to restart brokers.
In the update config, we restart the brokers to pick up the config.

IoannisRP committed:
k/client: expose cluster configuration apis for cluster

IoannisRP committed:
dt: extract cluster link tls provider to base

michael-redpanda commented on src/v/kafka/client/configuration.h:
```diff
@@ -71,6 +71,8 @@ struct tls_configuration {
     ss::future<ss::shared_ptr<ss::tls::certificate_credentials>>
     build_credentials() const;
+
+    bool operator==(const tls_configuration&) const = default;
```
nit: use the non-member friend version of the equality operator

michael-redpanda commented on src/v/kafka/client/brokers.cc:
```diff
@@ -57,6 +62,24 @@ ss::future<> brokers::do_erase(model::node_id node_id) {
+    vlog(
+      _logger->debug,
+      "Clear brokers: {}",
+      fmt::join(_brokers | std::views::values | to_string, ","));
```
praise: i love seeing this

michael-redpanda commented on src/v/kafka/client/cluster.cc:
```diff
@@ -364,10 +365,25 @@ void cluster::set_sasl_configuration(std::optional<sasl_configuration> creds) {
+    // all updates to the cluster brokers should happen under the
+    // metadata update lock
+    co_await _brokers.clear();
+    co_await initialize_metadata_with_seed();
```
I think you should do a full metadata update - if the brokers have changed, then the partition assignments in the topic cache are probably no longer valid.

michael-redpanda replied:
The initialization does either a metadata request with no topics requested or a describe cluster.  Both responses will not contain topic information

bharathv replied:
I think this should just call dispatch_and_apply_metadata_updates(nullopt), should also notify waiters.

michael-redpanda reviewed (commented):
nice!

bharathv commented on src/v/kafka/client/cluster.h:
```diff
@@ -129,6 +129,11 @@ class cluster {
         return _config.sasl_cfg;
     }
+
+    ss::future<> update_broker_config(
```
wonder if we should do an update_configuration(connection_configuration)? (so all fields are dynamic)

IoannisRP replied:
i like the idea, but better to keep it separately from this pr.

mmaslankaprv replied:
I am +1 for what Bharath is suggesting, why do we need to create this very specific APIs that accepts only a part of configuration ? Is there any obstacle that prevents use from passing a connection configuration here ?

IoannisRP replied:
> Is there any obstacle that prevents use from passing a connection configuration

out of the configurations of the `connection_configuration`, we don't currently propagate all of them into the `kafka::client::cluster`. Specifically, we don't update
```
    std::optional<ss::sstring> client_id;
    std::chrono::milliseconds max_metadata_age{10000};
    std::chrono::milliseconds connection_timeout{1000};
```

I would prefer the function that takes the full connection configuration to update everything properly. Alternatively, we have a function that seems like it is updating everything but is very selective internally, which i don't like.

That's why i suggest keeping the interface tight in this pr, and extenting the kafka client to what bharath suggested separately (and fixing the calls of `cluster_link::link`  to the new api).

michael-redpanda replied:
I also agree with Bharath here - take the whole structure and pick out what we need from it

IoannisRP commented:
changes in [force-push](https://github.com/redpanda-data/redpanda/compare/c20d5a80fe47a51ebcf4cfceea9f80d41c5c3656..cf01115974e8657a382d9d5ad13b0fad0626bf08):
- changed re-initialization of brokers to request full metadata

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/73938-->
<!--a51b067f-c4fe-48f0-9aed-52d50b066b9e-->
<!--1073601e-4678-4eaf-93f5-c01c2cf5f37c-->
<!--83e0e972-a846-4bf2-9c71-fcb0fc780b03-->
<!--title-->
## Retry command for [Build#73938](https://buildkite.com/redpanda/redpanda/builds/73938)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/random_node_operations_test.py::RedpandaNodeOperationsSmokeTest.test_node_ops_smoke_test@{"cloud_storage_type":1,"mixed_versions":false}
tests/rptest/tests/cluster_linking_e2e_test.py::ShadowLinkBasicTests.test_invalid_updates
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":2,"compaction_mode":"chunked_sliding_window","enable_failures":true,"mixed_versions":false,"with_iceberg":false}
```
<!--/body-->

michael-redpanda reviewed (dismissed):
lgtm - want to see if @bharathv is happy with your response and then we can merge

IoannisRP commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/cf01115974e8657a382d9d5ad13b0fad0626bf08..0e562c69e76cf6739e862aa27ed1ca9eab7bbe2f):
- remove test that was validating that bootstrap servers + tls cannot be changed.

bharathv commented on src/v/cluster_link/link.cc:
```diff
@@ -510,4 +512,20 @@ void link::maybe_update_sasl_configuration(
+      && broker_tls == client_config.broker_tls) {
+        return ss::now();
+    }
+
```
this is worthy of an info log IMO.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/73956-->
<!--e4cf2b22-53eb-4e7b-b6dc-cdab3eddf1f8-->
<!--e0ea1268-ab23-4ff6-9d87-1c1e50e7d067-->
<!--41450a6a-e4ad-479e-a680-35bdef20748b-->
<!--70dc021d-b3a8-4ccb-b58f-a2d158866148-->
<!--0cb58520-cb2e-42e3-a574-5ca996d3c927-->
<!--title-->
## Retry command for [Build#73956](https://buildkite.com/redpanda/redpanda/builds/73956)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":2,"compaction_mode":"chunked_sliding_window","enable_failures":true,"mixed_versions":false,"with_iceberg":false}
tests/rptest/tests/random_node_operations_test.py::RedpandaNodeOperationsSmokeTest.test_node_ops_smoke_test@{"cloud_storage_type":1,"mixed_versions":false}
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":2,"compaction_mode":"chunked_sliding_window","enable_failures":false,"mixed_versions":false,"with_iceberg":false}
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":1,"compaction_mode":"chunked_sliding_window","enable_failures":true,"mixed_versions":false,"with_iceberg":false}
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":1,"compaction_mode":"adjacent_merge","enable_failures":false,"mixed_versions":false,"with_iceberg":true}
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":2,"compaction_mode":"sliding_window","enable_failures":true,"mixed_versions":false,"with_iceberg":false}
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--54d1af30-c511-41c5-b62d-6c1afb8ce72a-->
<!--TEST RESULTS-->
<!--81c78e8d-1297-4f4b-9a74-310c36a65391-->
<!--TEST RESULTS-->
<!--e9e33833-a2da-4afc-acd5-c2eac1237109-->
<!--TEST RESULTS-->
<!--b667008c-4cb3-4072-8340-aa0564268ad9-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/73956">build#73956</a></summary>


| test_class                      | test_method              | test_arguments                                                                                                                                   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                 | test_history                                                                                                                                                           |
|:--------------------------------|:-------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:---------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| RandomNodeOperationsTest        | test_node_operations     | {"cloud_storage_type": 2, "compaction_mode": "adjacent_merge", "enable_failures": false, "mixed_versions": false, "with_iceberg": false}         | integration | https://buildkite.com/redpanda/redpanda/builds/73956#0199cec4-def3-45ac-adef-f1bdf92216e4 | FAIL          | 0/19     |                                        | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RandomNodeOperationsTest&test_method=test_node_operations            |
| RandomNodeOperationsTest        | test_node_operations     | {"cloud_storage_type": 1, "compaction_mode": "adjacent_merge", "enable_failures": false, "mixed_versions": false, "with_iceberg": false}         | integration | https://buildkite.com/redpanda/redpanda/builds/73956#0199cec4-def4-47fc-bc13-5a47354d6e36 | FAIL          | 0/11     |                                        | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RandomNodeOperationsTest&test_method=test_node_operations            |
| RandomNodeOperationsTest        | test_node_operations     | {"cloud_storage_type": 2, "compaction_mode": "chunked_sliding_window", "enable_failures": false, "mixed_versions": false, "with_iceberg": false} | integration | https://buildkite.com/redpanda/redpanda/builds/73956#0199cec4-def5-410f-b182-47d525d9d92e | FAIL          | 0/21     | The test has failed across all retries | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RandomNodeOperationsTest&test_method=test_node_operations            |
| RandomNodeOperationsTest        | test_node_operations     | {"cloud_storage_type": 1, "compaction_mode": "chunked_sliding_window", "enable_failures": false, "mixed_versions": false, "with_iceberg": false} | integration | https://buildkite.com/redpanda/redpanda/builds/73956#0199cec4-def6-4d16-90c1-b4ad05b7e263 | FAIL          | 0/15     |                                        | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RandomNodeOperationsTest&test_method=test_node_operations            |
| RandomNodeOperationsTest        | test_node_operations     | {"cloud_storage_type": 2, "compaction_mode": "sliding_window", "enable_failures": false, "mixed_versions": false, "with_iceberg": false}         | integration | https://buildkite.com/redpanda/redpanda/builds/73956#0199cec4-def7-4fee-b672-b5ebe7ebb469 | FAIL          | 0/16     |                                        | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RandomNodeOperationsTest&test_method=test_node_operations            |
| RandomNodeOperationsTest        | test_node_operations     | {"cloud_storage_type": 1, "compaction_mode": "sliding_window", "enable_failures": false, "mixed_versions": false, "with_iceberg": false}         | integration | https://buildkite.com/redpanda/redpanda/builds/73956#0199cec4-deed-4f42-bd72-9ceb8e2faec7 | FAIL          | 0/16     |                                        | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RandomNodeOperationsTest&test_method=test_node_operations            |
| RandomNodeOperationsTest        | test_node_operations     | {"cloud_storage_type": 1, "compaction_mode": "adjacent_merge", "enable_failures": false, "mixed_versions": false, "with_iceberg": true}          | integration | https://buildkite.com/redpanda/redpanda/builds/73956#0199cec4-def0-4e3a-a74c-5ca3e85560a0 | FAIL          | 0/21     | The test has failed across all retries | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RandomNodeOperationsTest&test_method=test_node_operations            |
| RandomNodeOperationsTest        | test_node_operations     | {"cloud_storage_type": 1, "compaction_mode": "chunked_sliding_window", "enable_failures": false, "mixed_versions": false, "with_iceberg": true}  | integration | https://buildkite.com/redpanda/redpanda/builds/73956#0199cec4-def2-4749-a5b3-acd0b7328159 | FAIL          | 0/17     |                                        | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RandomNodeOperationsTest&test_method=test_node_operations            |
| RandomNodeOperationsTest        | test_node_operations     | {"cloud_storage_type": 1, "compaction_mode": "sliding_window", "enable_failures": false, "mixed_versions": false, "with_iceberg": true}          | integration | https://buildkite.com/redpanda/redpanda/builds/73956#0199cec4-def4-47fc-bc13-5a47354d6e36 | FAIL          | 0/10     |                                        | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RandomNodeOperationsTest&test_method=test_node_operations            |
| RandomNodeOperationsTest        | test_node_operations     | {"cloud_storage_type": 2, "compaction_mode": "adjacent_merge", "enable_failures": true, "mixed_versions": false, "with_iceberg": false}          | integration | https://buildkite.com/redpanda/redpanda/builds/73956#0199cec4-def7-4fee-b672-b5ebe7ebb469 | FAIL          | 0/15     |                                        | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RandomNodeOperationsTest&test_method=test_node_operations            |
| RandomNodeOperationsTest        | test_node_operations     | {"cloud_storage_type": 1, "compaction_mode": "adjacent_merge", "enable_failures": true, "mixed_versions": false, "with_iceberg": false}          | integration | https://buildkite.com/redpanda/redpanda/builds/73956#0199cec4-deed-4f42-bd72-9ceb8e2faec7 | FAIL          | 0/1      |                                        | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RandomNodeOperationsTest&test_method=test_node_operations            |
| RandomNodeOperationsTest        | test_node_operations     | {"cloud_storage_type": 2, "compaction_mode": "chunked_sliding_window", "enable_failures": true, "mixed_versions": false, "with_iceberg": false}  | integration | https://buildkite.com/redpanda/redpanda/builds/73956#0199cec4-deee-4639-b421-d753b7a51020 | FAIL          | 0/1      |                                        | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RandomNodeOperationsTest&test_method=test_node_operations            |
| RandomNodeOperationsTest        | test_node_operations     | {"cloud_storage_type": 1, "compaction_mode": "chunked_sliding_window", "enable_failures": true, "mixed_versions": false, "with_iceberg": false}  | integration | https://buildkite.com/redpanda/redpanda/builds/73956#0199cec4-def0-4e3a-a74c-5ca3e85560a0 | FAIL          | 0/1      | The test has failed across all retries | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RandomNodeOperationsTest&test_method=test_node_operations            |
| RandomNodeOperationsTest        | test_node_operations     | {"cloud_storage_type": 2, "compaction_mode": "sliding_window", "enable_failures": true, "mixed_versions": false, "with_iceberg": false}          | integration | https://buildkite.com/redpanda/redpanda/builds/73956#0199cec4-def1-4776-8767-f94a021b7e9f | FAIL          | 0/21     | The test has failed across all retries | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RandomNodeOperationsTest&test_method=test_node_operations            |
| RandomNodeOperationsTest        | test_node_operations     | {"cloud_storage_type": 1, "compaction_mode": "sliding_window", "enable_failures": true, "mixed_versions": false, "with_iceberg": false}          | integration | https://buildkite.com/redpanda/redpanda/builds/73956#0199cec4-def2-4749-a5b3-acd0b7328159 | FAIL          | 0/17     |                                        | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RandomNodeOperationsTest&test_method=test_node_operations            |
| RandomNodeOperationsTest        | test_node_operations     | {"cloud_storage_type": 1, "compaction_mode": "adjacent_merge", "enable_failures": true, "mixed_versions": false, "with_iceberg": true}           | integration | https://buildkite.com/redpanda/redpanda/builds/73956#0199cec4-def4-47fc-bc13-5a47354d6e36 | FAIL          | 0/10     |                                        | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RandomNodeOperationsTest&test_method=test_node_operations            |
| RandomNodeOperationsTest        | test_node_operations     | {"cloud_storage_type": 1, "compaction_mode": "chunked_sliding_window", "enable_failures": true, "mixed_versions": false, "with_iceberg": true}   | integration | https://buildkite.com/redpanda/redpanda/builds/73956#0199cec4-def6-4d16-90c1-b4ad05b7e263 | FAIL          | 0/15     |                                        | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RandomNodeOperationsTest&test_method=test_node_operations            |
| RandomNodeOperationsTest        | test_node_operations     | {"cloud_storage_type": 1, "compaction_mode": "sliding_window", "enable_failures": true, "mixed_versions": false, "with_iceberg": true}           | integration | https://buildkite.com/redpanda/redpanda/builds/73956#0199cec4-deed-4f42-bd72-9ceb8e2faec7 | FAIL          | 0/16     |                                        | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RandomNodeOperationsTest&test_method=test_node_operations            |
| RandomNodeOperationsTest        | test_node_operations     | {"cloud_storage_type": 2, "compaction_mode": "sliding_window", "enable_failures": true, "mixed_versions": true, "with_iceberg": false}           | integration | https://buildkite.com/redpanda/redpanda/builds/73956#0199cec4-def3-45ac-adef-f1bdf92216e4 | FLAKY         | 18/19    |                                        | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RandomNodeOperationsTest&test_method=test_node_operations            |
| RedpandaNodeOperationsSmokeTest | test_node_ops_smoke_test | {"cloud_storage_type": 1, "mixed_versions": false}                                                                                               | integration | https://buildkite.com/redpanda/redpanda/builds/73956#0199cec4-7d48-4ab1-aa30-9e814c98f202 | FAIL          | 0/21     | The test has failed across all retries | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RedpandaNodeOperationsSmokeTest&test_method=test_node_ops_smoke_test |
| RedpandaNodeOperationsSmokeTest | test_node_ops_smoke_test | {"cloud_storage_type": 1, "mixed_versions": false}                                                                                               | integration | https://buildkite.com/redpanda/redpanda/builds/73956#0199cec4-def1-4776-8767-f94a021b7e9f | FAIL          | 0/1      | The test has failed across all retries | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RedpandaNodeOperationsSmokeTest&test_method=test_node_ops_smoke_test |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74439">build#74439</a></summary>


| test_class                      | test_method                          | test_arguments                                                   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                               |
|:--------------------------------|:-------------------------------------|:-----------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:---------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkBasicTests            | test_rapid_shadow_link_toggling      | null                                                             | integration | https://buildkite.com/redpanda/redpanda/builds/74439#0199f3f4-0659-44f7-a408-7ba6b9e73c44 | FLAKY         | 1/21     | upstream reliability is '92.65175718849841'. current run reliability is '4.761904761904762'. drift is 87.88985 and the allowed drift is set to 50. The test should FAIL | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkBasicTests&test_method=test_rapid_shadow_link_toggling         |
| ShadowLinkingReplicationTests   | test_replication_with_failures       | null                                                             | integration | https://buildkite.com/redpanda/redpanda/builds/74439#0199f3f4-0660-4910-aabf-c42c9fae7c63 | FLAKY         | 11/21    | upstream reliability is '77.13920817369093'. current run reliability is '52.38095238095239'. drift is 24.75826 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_with_failures |
| ConfigurationUpdateTest         | test_two_nodes_update                | null                                                             | integration | https://buildkite.com/redpanda/redpanda/builds/74439#0199f3f4-065c-446b-9024-95fd36afe80b | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS               | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ConfigurationUpdateTest&test_method=test_two_nodes_update                |
| DataMigrationsApiTest           | test_creating_and_listing_migrations | null                                                             | integration | https://buildkite.com/redpanda/redpanda/builds/74439#0199f3f4-065e-4e04-8a5b-b302c4b711e2 | FLAKY         | 20/21    | upstream reliability is '99.00398406374502'. current run reliability is '95.23809523809523'. drift is 3.76589 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DataMigrationsApiTest&test_method=test_creating_and_listing_migrations   |
| NodesDecommissioningTest        | test_recommissioning_node            | {"cloud_topic": false}                                           | integration | https://buildkite.com/redpanda/redpanda/builds/74439#0199f3fa-eaad-45ae-b58d-abc5a6218063 | FLAKY         | 20/21    | upstream reliability is '98.01324503311258'. current run reliability is '95.23809523809523'. drift is 2.77515 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_recommissioning_node           |
| RedpandaNodeOperationsSmokeTest | test_node_ops_smoke_test             | {"cloud_storage_type": 1, "mixed_versions": true}                | integration | https://buildkite.com/redpanda/redpanda/builds/74439#0199f3fa-eaa9-4315-8524-091ae44ba9d7 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS               | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RedpandaNodeOperationsSmokeTest&test_method=test_node_ops_smoke_test     |
| TimeQueryTest                   | test_timequery                       | {"batch_cache": false, "cloud_storage": true, "spillover": true} | integration | https://buildkite.com/redpanda/redpanda/builds/74439#0199f3f4-0659-44f7-a408-7ba6b9e73c44 | FLAKY         | 20/21    | upstream reliability is '99.76470588235294'. current run reliability is '95.23809523809523'. drift is 4.52661 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=TimeQueryTest&test_method=test_timequery                                 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74614">build#74614</a></summary>


| test_class                 | test_method              | test_arguments                                           | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                   | test_history                                                                                                                                                  |
|:---------------------------|:-------------------------|:---------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:--------------------------------------------------------------------------------------------------------------------------------------------------------------|
| CompactionGapsTest         | test_translation_no_gaps | {"catalog_type": "rest_hadoop", "cloud_storage_type": 1} | integration | https://buildkite.com/redpanda/redpanda/builds/74614#019a07b5-5d7e-43c4-8123-9358349508db | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS                | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=CompactionGapsTest&test_method=test_translation_no_gaps     |
| ShadowLinkingRandomOpsTest | test_node_operations     | {"failures": false}                                      | integration | https://buildkite.com/redpanda/redpanda/builds/74614#019a07b5-5d88-4c86-a7ad-d6e63df4067a | FLAKY         | 10/21    | upstream reliability is '47.61904761904761'. current run reliability is '47.61904761904761'. drift is 0.0 and the allowed drift is set to 50. The test should PASS       | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingRandomOpsTest&test_method=test_node_operations |
| ShadowLinkingRandomOpsTest | test_node_operations     | {"failures": true}                                       | integration | https://buildkite.com/redpanda/redpanda/builds/74614#019a07ad-ba45-4074-9111-7e53129bc98c | FLAKY         | 3/21     | upstream reliability is '66.66666666666666'. current run reliability is '14.285714285714285'. drift is 52.38095 and the allowed drift is set to 50. The test should FAIL | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingRandomOpsTest&test_method=test_node_operations |
| ShadowLinkingRandomOpsTest | test_node_operations     | {"failures": true}                                       | integration | https://buildkite.com/redpanda/redpanda/builds/74614#019a07b5-5d7c-45f0-bd12-d1a8dfd5b810 | FLAKY         | 3/21     | upstream reliability is '66.66666666666666'. current run reliability is '14.285714285714285'. drift is 52.38095 and the allowed drift is set to 50. The test should FAIL | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingRandomOpsTest&test_method=test_node_operations |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74750">build#74750</a></summary>


| test_class                       | test_method                                             | test_arguments                               | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                    | test_history                                                                                                                                                                                   |
|:---------------------------------|:--------------------------------------------------------|:---------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:--------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| NodesDecommissioningTest         | test_decommissioning_finishes_after_manual_cancellation | {"cloud_topic": true, "delete_topic": false} | integration | https://buildkite.com/redpanda/redpanda/builds/74750#019a0e85-63f3-4324-87f5-61f8c74a7761 | FLAKY         | 18/21    | upstream reliability is '98.53479853479854'. current run reliability is '85.71428571428571'. drift is 12.82051 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_decommissioning_finishes_after_manual_cancellation |
| ShadowLinkingRandomOpsTest       | test_node_operations                                    | {"failures": false}                          | integration | https://buildkite.com/redpanda/redpanda/builds/74750#019a0e85-63eb-43d7-903a-e6de8c649199 | FLAKY         | 3/21     | upstream reliability is '80.96885813148789'. current run reliability is '14.285714285714285'. drift is 66.68314 and the allowed drift is set to 50. The test should FAIL  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingRandomOpsTest&test_method=test_node_operations                                  |
| ShadowLinkingRandomOpsTest       | test_node_operations                                    | {"failures": true}                           | integration | https://buildkite.com/redpanda/redpanda/builds/74750#019a0e7c-700a-4e27-9f13-dec86606affd | FLAKY         | 3/21     | upstream reliability is '32.72251308900523'. current run reliability is '14.285714285714285'. drift is 18.4368 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingRandomOpsTest&test_method=test_node_operations                                  |
| ShadowLinkingRandomOpsTest       | test_node_operations                                    | {"failures": true}                           | integration | https://buildkite.com/redpanda/redpanda/builds/74750#019a0e85-63ec-4c24-903b-629a776cbbf6 | FLAKY         | 3/21     | upstream reliability is '62.189054726368155'. current run reliability is '14.285714285714285'. drift is 47.90334 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingRandomOpsTest&test_method=test_node_operations                                  |
| WriteCachingFailureInjectionTest | test_unavoidable_data_loss                              | null                                         | integration | https://buildkite.com/redpanda/redpanda/builds/74750#019a0e85-63ef-4a65-b640-d6a7db0eea32 | FLAKY         | 19/21    | upstream reliability is '92.29885057471265'. current run reliability is '90.47619047619048'. drift is 1.82266 and the allowed drift is set to 50. The test should PASS    | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionTest&test_method=test_unavoidable_data_loss                      |
</details>
<!--/body-->

mmaslankaprv commented:
Can we merge this after: https://github.com/redpanda-data/redpanda/pull/27703 ?

michael-redpanda commented:
Force push:

* Rebased off of dev

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/74439-->
<!--52e3e4cf-f682-454f-ac3f-221c43ab099e-->
<!--298cbcc5-e9be-445f-a254-ad419f8e8553-->
<!--title-->
## Retry command for [Build#74439](https://buildkite.com/redpanda/redpanda/builds/74439)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/cluster_linking_e2e_test.py::ShadowLinkBasicTests.test_rapid_shadow_link_toggling
```
<!--/body-->

IoannisRP commented:
changes in [force-push](https://github.com/redpanda-data/redpanda/compare/82006ead6fcc4843cf819b840e3602b6382e1152..791b6807f0133ddf31ac8fdd5af3ebd39e0a1f6f):
- rebase to `dev` to pick up fix on failing dt test

IoannisRP commented:
changes in [force-push](https://github.com/redpanda-data/redpanda/compare/791b6807f0133ddf31ac8fdd5af3ebd39e0a1f6f..76c51796cc19186e1fd9dfeacdfe7dc2f73f1dbc):
- rebase to `dev` to resolve conflict

michael-redpanda commented on src/v/cluster_link/link.cc:
```diff
@@ -176,8 +176,8 @@ ss::future<cl_result<void>> link::register_task(task_factory* tf) {
-void link::update_config(
-  model::metadata config, ::model::revision_id revision) {
+ss::future<>
+link::update_config(model::metadata config, ::model::revision_id revision) {
```
Why do we have to futurize this method?  Can we apply the configuration in a background fiber?

michael-redpanda commented on src/v/cluster_link/manager.cc:
```diff
@@ -470,7 +470,7 @@ manager::handle_on_link_change(model::id_t id, ::model::revision_id revision) {
           id,
           link_metadata);
-        it->second->update_config(link_metadata.copy(), revision);
+        co_await it->second->update_config(link_metadata.copy(), revision);
```
Though we won't run into issues now (because we only permit a single link), in the future I worry about the iterator becoming invalidated here.

IoannisRP committed:
k/client: Make connection_configuration equal comparable

IoannisRP committed:
k/client: Make connection_configuration formattable

IoannisRP committed:
cluster_link: Update link to use the new kafka client configuration interface

IoannisRP commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/76c51796cc19186e1fd9dfeacdfe7dc2f73f1dbc..ccbfc176cf4b585ff39fc14690108b05e963618e):
- implemented `update_config(connection_configuration)` function in cluster that only operates on sasl_config and bootstrap brokers
- moved broker update to background fiber

Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/ccbfc176cf4b585ff39fc14690108b05e963618e..9a50f05e7e07888837f148903772dc8b457d559b):
- rebase to `dev` to fix merge conflict

michael-redpanda commented:
The type checking issue was fixed here: https://github.com/redpanda-data/redpanda/pull/28134

michael-redpanda commented on src/v/kafka/client/cluster.cc:
```diff
@@ -370,13 +371,39 @@ void cluster::update_configuration(connection_configuration config) {
+              std::move(config.initial_brokers), std::move(config.broker_tls))
+              .handle_exception([this](std::exception_ptr ex) {
+                  vlog(_logger.warn, "Failed to update broker config: {}", ex);
+              });
```
You should use `ssx::spawn_with_gate` here

michael-redpanda replied:
I also worry here about what would happen if a client configuration changes rapidly, then we're unsure which one will 'win'.  Could use a task queue here

bharathv replied:
+1 on queueing the config updates.

michael-redpanda reviewed (commented):
looks good - would be interested in having @mmaslankaprv take a quick peak since he has done so much work in `kafka/client/cluster`

bharathv commented on src/v/kafka/client/cluster.cc:
```diff
@@ -370,13 +371,39 @@ void cluster::update_configuration(connection_configuration config) {
+                  vlog(_logger.warn, "Failed to update broker config: {}", ex);
+              });
+    }
+
```
can we just replace the whole config if there is a mismatch, why do we have to cherry-pick specific stuff? Just replace the config and and restart the brokers instances?

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/74614-->
<!--9d71a0b1-3bde-4846-be66-f3a3977bf010-->
<!--15a69ed3-d351-445b-89e7-1fae22d766c1-->
<!--title-->
## Retry command for [Build#74614](https://buildkite.com/redpanda/redpanda/builds/74614)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/shadow_linking_rnot_test.py::ShadowLinkingRandomOpsTest.test_node_operations@{"failures":true}
```
<!--/body-->

IoannisRP commented:
changes in [force-push](https://github.com/redpanda-data/redpanda/compare/9a50f05e7e07888837f148903772dc8b457d559b..5946d44db6da7e7f180c0ccbf989827153b34b48):
- rebase to `dev`
- added queuing for config updates
- config update replaces the whole config and restarts brokers.

IoannisRP committed:
pp/test: Add call to stop for kclients in pp test

These tests don't call start, so they didn't need to call stop, till
now. In the next commit, an ssx::work_queue will be added to the kafka
client cluster, which makes the call to stop mandatory, even if it
hasn't been started.

IoannisRP commented:
changes in [force-push](https://github.com/redpanda-data/redpanda/compare/5946d44db6da7e7f180c0ccbf989827153b34b48..343caba9fb74d7856c06b1540872e021e6bb0c8a):
- updated pp test to properly call stop on all clients. Previously it was failing because clients now have a `ssx::work_queue` which requires the `shutdown` method to be called upon, during the `kafka::client::cluster::stop`, even if `start` hasn't been called.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/74750-->
<!--839dd05e-9315-4275-a100-45e44917997e-->
<!--title-->
## Retry command for [Build#74750](https://buildkite.com/redpanda/redpanda/builds/74750)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/shadow_linking_rnot_test.py::ShadowLinkingRandomOpsTest.test_node_operations@{"failures":false}
```
<!--/body-->
